### PR TITLE
[Merged by Bors] - feat(EllipticCurve): a coordinate ring over a domain is a domain

### PR DESCRIPTION
**Pull request by alreadydone**
Generalize `instIsDomainCoordinateRing` from normalized GCD domains to all domains, by reducing to the field case via the injective ring hom to the field of fractions. (This works basically because the coordinate ring is a free module over the base ring and therefore flat.) To get this done, we move various declarations (some instances and `basis` API) up, and introduce `CoordinateRing.map/map_mk/map_smul/map_injective`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 1]
```diff
@@ -355,10 +463,6 @@ instance instIsScalarTowerCoordinateRing : IsScalarTower R R[X] W.CoordinateRing
   Quotient.isScalarTower R R[X] _
 #align weierstrass_curve.coordinate_ring.is_scalar_tower WeierstrassCurve.Affine.CoordinateRing.instIsScalarTowerCoordinateRing
 
-instance instSubsingletonCoordinateRing [Subsingleton R] : Subsingleton W.CoordinateRing :=
-  Module.subsingleton R[X] _
-#align weierstrass_curve.coordinate_ring.subsingleton WeierstrassCurve.Affine.CoordinateRing.instSubsingletonCoordinateRing
-
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:361]: I think the following section organisation is better. Move the entire `Algebra` section "The coordinate ring as an `R[X]`-algebra" (from `instAlgebraCoordinateRing` to `smul_basis_mul_Y`) to right before the `Ring` section "Ideals in the coordinate ring over a ring". This won't literally compile because of misplaced definitions, but the following changes should make it work: (1) move `quotientXYIdealEquiv` from the `Algebra` section to right after `XYIdeal_add_eq` in the `Ring` section because it needs the definition of `XYIdeal`; (2) move `mk` from the `Ring` section to right before `basis` in the `Algebra` section because some of the lemmas for `basis` require this definition.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 2]
```diff
@@ -93,25 +93,133 @@ namespace CoordinateRing
...
+        (p - q * (C W.a₁ * X + C W.a₃)) • mk W Y := by
+  have Y_sq : mk W Y ^ 2 =
+      mk W (C (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆) - C (C W.a₁ * X + C W.a₃) * Y) := by
+    exact AdjoinRoot.mk_eq_mk.mpr ⟨1, by rw [polynomial]; ring1⟩
+  simp_rw [smul, add_mul, mul_assoc, ← sq, Y_sq, C_sub, map_sub, C_mul, _root_.map_mul]
+  ring1
+set_option linter.uppercaseLean3 false in
+#align weierstrass_curve.coordinate_ring.smul_basis_mul_Y WeierstrassCurve.Affine.CoordinateRing.smul_basis_mul_Y
+
+section map
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:182]: This probably doesn't need a section, unless you intend to make this a lot bigger. In which case I suggest capitalising it (like the other sections) and putting it outside the parent `Ring`/`Algebra` section.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 3]
```diff
@@ -93,25 +93,133 @@ namespace CoordinateRing
...
+
+variable {S} [CommRing S] (f : R →+* S)
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:184]: For consistency with the `map`s in `Affine.lean`, make this `variable {A : Type v} [CommRing A] (φ : R →+* A)` with a new universe `v` added at the top of the file.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 4]
```diff
@@ -93,25 +93,133 @@ namespace CoordinateRing
...
+lemma map_injective (hf : Injective f) : Injective (map W f) :=
+  (injective_iff_map_eq_zero _).mpr fun y hy ↦ by
+    obtain ⟨p, q, rfl⟩ := exists_smul_basis_eq y
+    simp_rw [map_add, CoordinateRing.map_smul, map_one, map_mk, map_X] at hy
+    obtain ⟨hp, hq⟩ := smul_basis_eq_zero hy
+    rw [Polynomial.map_eq_zero_iff hf] at hp hq
+    simp_rw [hp, hq, zero_smul, add_zero]
+
+instance instIsDomainCoordinateRing [IsDomain R] : IsDomain W.CoordinateRing :=
+  haveI : IsDomain (CoordinateRing <| W.map <| algebraMap R <| FractionRing R) :=
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:211]: I think that `have` in Lean 4 is `haveI` in Lean 3? Meanwhile `haveI` in Lean 4 inlines this, but that's not needed.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 5]
```diff
@@ -93,25 +93,133 @@ namespace CoordinateRing
...
+  AdjoinRoot.lift ((AdjoinRoot.of _).comp <| mapRingHom f) (AdjoinRoot.root _) <| by
+    simp [← eval₂_map, ← map_polynomial]
+
+variable {W f}
+
+lemma map_mk {x : R[X][X]} : map W f (mk W x) = mk (W.map f) (x.map <| mapRingHom f) := by
+  simp only [map, AdjoinRoot.lift_mk, ← eval₂_map]; apply AdjoinRoot.aeval_eq
+
+protected lemma map_smul {x : R[X]} {y : W.CoordinateRing} :
+    map W f (x • y) = x.map f • map W f y := by rw [smul, _root_.map_mul, map_mk, smul, map_C]; rfl
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:199]: Probably don't need to protect this? In any case it looks a bit crowded, and as per conventions I think everything after `by` should be in the next line.
```suggestion
lemma map_smul {x : R[X]} {y : W.CoordinateRing} : map W φ (x • y) = map W φ x • map W φ y := by
  rw [smul, _root_.map_mul, map_mk, smul, map_C]
  rfl
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 6]
```diff
@@ -93,25 +93,133 @@ namespace CoordinateRing
...
+
+open Function in
+lemma map_injective (hf : Injective f) : Injective (map W f) :=
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:202]: Everywhere I just do `Function.Injective` since it doesn't eat an extra line:
```suggestion
lemma map_injective (hf : Function.Injective φ) : Function.Injective <| map W φ :=
```

**Review by Multramate — CHANGES_REQUESTED** [review]
Some stylistic suggestions

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 5 above)
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:199]: [map_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/GroupAction/Hom.html#map_smul) is in root namespace so it's certainly reasonable to protect, no?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 4 above)
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:211]: Yeah, this is my bad habit, but probably harmless ...

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 3 above)
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:184]: I think the name `A` was introduced back when there's only`baseChange` (with an algebra) but no `map`. Isn't S more appropriate for the target of an arbitrary ring hom?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean [hunk 7]
```diff
@@ -188,6 +188,9 @@ lemma eval_polynomial_zero : (W.polynomial.eval 0).eval 0 = -W.a₆ := by
   simp only [← C_0, eval_polynomial, zero_add, zero_sub, mul_zero, zero_pow <| Nat.succ_ne_zero _]
 #align weierstrass_curve.eval_polynomial_zero WeierstrassCurve.Affine.eval_polynomial_zero
 
+lemma map_polynomial {S} [CommRing S] (f : R →+* S) :
+    polynomial (W.map f) = W.polynomial.map (mapRingHom f) := by simp [polynomial]
+
```
[the remark this answers is not in the document]
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:193]: I think since `WeierstrassCurve.Affine` is just an abbrev of `WeierstrassCurve`, it's not defeq abuse even if we don't insert `toAffine`. I think you introduce the abbrev just to be able to use dot notation? Because there's no syntax in Lean 4 that allows using dot notation inside a sub-namespace.

Of course if e.g. both `Affine` and `Jacobian` namespaces are open, then Lean could get confused which `polynomial` you are referring to, but in that case I still think both `(W.map φ).toAffine.polynomial` and `Affine.polynomial (W.map φ)` are acceptable solutions.

[the remark this answers is not in the document]
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:193]: I introduced the `abbrev` mostly to avoid the clashes for `Affine`/`Projective`/`Jacobian`, and I think both are fine. My comment here is just for the relocation of the lemma.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 3 above)
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:184]: Fair enough. Maybe you can change the few `A`s in `Affine.lean` to be `S`s if it's not too much effort. We'll make the change to `RingHomClass` in a future PR.

Regarding universe variables: what is the default if it's not explicit? I always make the effort to give them different universe variables whenever necessary to be as general as possible.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 5 above)
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:199]: Sure, this is mostly for the crowded line.

**Comment by Multramate** [comment]
Sorry I have no idea what I just did with the other branch. I just moved a few things around according to what I said above, hopefully you don't mind. Let's merge this small standalone PR before moving on to your other PRs.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 8]
```diff
@@ -95,27 +95,143 @@ abbrev FunctionField : Type u :=
...
+  exact ⟨h hpq 0, h hpq 1⟩
+#align weierstrass_curve.coordinate_ring.smul_basis_eq_zero WeierstrassCurve.Affine.CoordinateRing.smul_basis_eq_zero
+
+variable {W} in
+lemma exists_smul_basis_eq (x : W.CoordinateRing) :
+    ∃ p q : R[X], p • (1 : W.CoordinateRing) + q • mk W Y = x := by
+  have h := (CoordinateRing.basis W).sum_equivFun x
+  erw [Fin.sum_univ_succ, Fin.sum_univ_one, basis_zero, basis_one] at h
+  exact ⟨_, _, h⟩
+#align weierstrass_curve.coordinate_ring.exists_smul_basis_eq WeierstrassCurve.Affine.CoordinateRing.exists_smul_basis_eq
```
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:177]: Why is this better than
```suggestion
variable {W}

lemma smul (x : R[X]) (y : W.CoordinateRing) : x • y = mk W (C x) * y :=
  (algebraMap_smul W.CoordinateRing x y).symm
#align weierstrass_curve.coordinate_ring.smul WeierstrassCurve.Affine.CoordinateRing.smul

lemma smul_basis_eq_zero {p q : R[X]} (hpq : p • (1 : W.CoordinateRing) + q • mk W Y = 0) :
    p = 0 ∧ q = 0 := by
  have h := Fintype.linearIndependent_iff.mp (CoordinateRing.basis W).linearIndependent ![p, q]
  erw [Fin.sum_univ_succ, basis_zero, Fin.sum_univ_one, basis_one] at h
  exact ⟨h hpq 0, h hpq 1⟩
#align weierstrass_curve.coordinate_ring.smul_basis_eq_zero WeierstrassCurve.Affine.CoordinateRing.smul_basis_eq_zero

lemma exists_smul_basis_eq (x : W.CoordinateRing) :
    ∃ p q : R[X], p • (1 : W.CoordinateRing) + q • mk W Y = x := by
  have h := (CoordinateRing.basis W).sum_equivFun x
  erw [Fin.sum_univ_succ, Fin.sum_univ_one, basis_zero, basis_one] at h
  exact ⟨_, _, h⟩
#align weierstrass_curve.coordinate_ring.exists_smul_basis_eq WeierstrassCurve.Affine.CoordinateRing.exists_smul_basis_eq

variable (W)
```
?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 9]
```diff
@@ -525,6 +541,7 @@ lemma degree_norm_ne_one [IsDomain R] (x : W.CoordinateRing) :
   exact (lt_max_of_lt_right <| by exact (cmp_eq_lt_iff ..).mp rfl).ne'
 #align weierstrass_curve.coordinate_ring.degree_norm_ne_one WeierstrassCurve.Affine.CoordinateRing.degree_norm_ne_one
 
+variable {W} in
```
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:544]: Same here: why this change?

**Review by alreadydone — COMMENTED** [review]
LGTM otherwise

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 8 above)
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:177]: I would much prefer finer control on the explicit/implicitness of the variables. When I move things around they sometimes get accidentally introduced/hidden without me noticing. Sometimes this doesn't affect the lemmas in the same file that use that specific lemma (e.g. they're just `rw`s without specifying the arguments), but affects downstream files (e.g. they're `exact`s supplying all the arguments). When I inevitably split this file in the future to match the Jacobian split I don't want to get confused myself.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 9 above)
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:544]: I guess this one preserves the total number of lines :)

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 8 above)
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:177]: I see, this is a good cause!

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean [hunk 10]
```diff
@@ -95,27 +95,143 @@ abbrev FunctionField : Type u :=
...
-#align weierstrass_curve.coordinate_ring.is_domain_of_field WeierstrassCurve.Affine.CoordinateRing.instIsDomainCoordinateRing_of_Field
+noncomputable instance : Algebra R[X] W.CoordinateRing :=
+  Quotient.algebra R[X]
+#align weierstrass_curve.coordinate_ring.algebra WeierstrassCurve.Affine.CoordinateRing.instAlgebraPolynomial
+
+instance : IsScalarTower R R[X] W.CoordinateRing :=
+  Quotient.isScalarTower R R[X] _
+#align weierstrass_curve.coordinate_ring.is_scalar_tower WeierstrassCurve.Affine.CoordinateRing.instIsScalarTowerPolynomial
+
+instance [Subsingleton R] : Subsingleton W.CoordinateRing :=
```
Comment (YaelDillies) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:114]: According to the naming convention, this should be `WeierstrassCurve.Affine.CoordinateRing.insSubsingleton`, so
```suggestion
instance instSubsingleton [Subsingleton R] : Subsingleton W.CoordinateRing :=
```

**Review by YaelDillies — COMMENTED** [review]
~~Looks good, but why did the instances go nameless?~~ All good

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean (hunk 10 above)
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:114]: I confirm that this is the auto-generated name. The full list of names are:
WeierstrassCurve.Affine.CoordinateRing.instAlgebra
WeierstrassCurve.Affine.CoordinateRing.instAlgebraPolynomial
WeierstrassCurve.Affine.CoordinateRing.instIsScalarTowerPolynomial
WeierstrassCurve.Affine.CoordinateRing.instSubsingleton
WeierstrassCurve.Affine.CoordinateRing.instIsDomain

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:114]: I don't think there is a style convention that we should manually name instances? Now that the instance name generation actually produces sensible names, I do feel that we should let it do its job whenever possible.

Comment (YaelDillies) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:114]: Oh sorry, I thought the autogenerated name now was something like `Blah.instSubsingleton_mathlib`

**Comment by YaelDillies** [comment]
maintainer merge

(this is actually mostly understandable to me even though I don't know the first thing about elliptic curves)

**Comment by alreadydone** [comment]
Thanks! 
@Multramate Please remember to delete the branch https://github.com/leanprover-community/mathlib4/tree/CoordinateRing_IsDomain_branch if it's no longer used.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge
